### PR TITLE
[MINOR][SQL][DOCS][2.4] Fix the timestamp pattern in the example for `to_timestamp`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to change the description of the `to_timestamp()` function, and change the pattern in the example. 

### Why are the changes needed?
To inform users about valid patterns for `to_timestamp` function.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
N/A

### CODE REVIEW

**Comment by MaxGekk** [comment]
@dongjoon-hyun Please, take a look at this.

**Comment by dongjoon-hyun** [comment]
It's too late for RC2, @MaxGekk . Let's consider this PR after RC2 vote.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: Let's not touch this file.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Let's not touch this.

**Comment by dongjoon-hyun** [comment]
In general, this seems to be a follow-up of SPARK-23792 (at 2.4.0). Since it's too old JIRA, we had better use another JIRA or use `[MINOR]` like you. I support your choice. I believe that one thing we need is that `yyyy-MM-dd HH:mm:ss.SSSS` -> `yyyy-MM-dd HH:mm:ss` change. I recommend to focus on the above change only.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: May I ask you to explain why we shouldn't clarify function's behavior here?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Just in case, are you ok with this change?

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Of course, you can do whatever you want with much more details, but I want to minimize the change.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Same with the others.

**Comment by dongjoon-hyun** [comment]
@MaxGekk . Every committer's suggestion is his/her own criteria for his/her acceptance. I am also able to merge only what I can agree. And, as you know, the other committers also have different opinion and they will merge this if they agree with AS-IS status more. It always does. In addition, I don't complain about the other committer's decision when I understand it's on the edge. Since this PR is yours, it's up to you always~ @MaxGekk .

**Comment by dongjoon-hyun** [comment]
Anyway, thank you always for you active contribution. Apache Spark community really needs that. 😄

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: @MaxGekk, sorry if I missed some contexts but do you mind the ones below don't work? Seems working find in Spark 2.4.4:

```
...
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 2.4.4
      /_/
...
scala> sql("select to_timestamp('2011-01-01 12:12:12.4321')").show(false)
+----------------------------------------+
|to_timestamp('2011-01-01 12:12:12.4321')|
+----------------------------------------+
|2011-01-01 12:12:12.4321                |
+----------------------------------------+


scala> sql("select to_timestamp('2011-01-01 12:12:12.4321')").printSchema()
root
 |-- to_timestamp('2011-01-01 12:12:12.4321'): timestamp (nullable = true)
```

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Usually it's better to make the changes minimised if we have to do it to old branches to reduce the potential conflicts when backporting. Seems this is the case.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Right, it works due to your changes https://github.com/apache/spark/pull/17901 . When the `format` parameter is omitted, `to_timestamp` uses another parsing mechanism via `DateTimeUtils.stringToTimestamp` and the result is not truncated to seconds as we do when the `format` is provided. Weird behavior. :-(

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Since SimpleDateFormat/FastDateFormat cannot parsing in microseconds precision, I would propose to switch `to_timestamp` on the parser introduced by https://github.com/apache/spark/pull/26507 and port back https://github.com/apache/spark/pull/24420 to 2.4. At the moment, the behavior of `to_timestamp` doesn't seem consistent. /cc @cloud-fan WDYT?

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: > Since SimpleDateFormat/FastDateFormat cannot parsing in microseconds precision

It has been there from the very first place. I think it's not clear if we should take it as a bug to backport.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: Anyway, I think it's better to discuss about this out of this place.

**Review by HyukjinKwon — APPROVED** [review]
LGTM. Since @dongjoon-hyun is preparing the release, I will leave it to him though.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: what's wrong with ".SSSS"?

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: 1. The result is incorrect, look at the PR which fixes that https://github.com/apache/spark/pull/26507, and here is an example:
```
    val sdf = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSSS")
    val res = sdf.parse("1970-01-01 00:00:00.1234")
    println(sdf.format(res))

1970-01-01 00:00:01.0234
```
2. And the result of `to_timestamp` is truncated to seconds.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: for 1), is it fixed in 2.4 or not?

for 2), have we documented it?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: SPARK-29904 is merged at 2.4.5. In the worst case, we can revert @MaxGekk 's #26507 because of this.

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: > for 1), is it fixed in 2.4 or not?

@cloud-fan It is fixed in JSON and CSV datasources but for `to_timestamp()` and other functions.

> for 2), have we documented it?

I have tried to document it here in the PR

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: > In the worst case, we can revert @MaxGekk 's #26507 because of this.

@dongjoon-hyun Why? What's the worst case? I do think we should apply the fix in other places but not reverting it.

Comment (MaxGekk) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: @cloud-fan I tried to document the restriction of `to_timestmp()` in the PR but @dongjoon-hyun convinced me to do not do that here, see https://github.com/apache/spark/pull/27438#issuecomment-581262355 " I believe that one thing we need is that yyyy-MM-dd HH:mm:ss.SSSS -> yyyy-MM-dd HH:mm:ss change. I recommend to focus on the above change only."

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: @MaxGekk . To be clear, I want to make it sure we don't have any regression at 2.4.5 as a release manager. You raised this issue and we are investigating the relevant PRs. That's all. We are considering the scope and effect. So far, we didn't make any decision.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala
Comment (MaxGekk) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: If you don't mind, I would like to say here that currently supported pattern for seconds fractions is `SSS` only. The second s fractions can be parsed but will be ignored while casting to TimestampType.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (AS-IS). Thanks, @MaxGekk and @HyukjinKwon .
Let's postpone this PR during RC2 vote period.

**Comment by cloud-fan** [comment]
maybe we can just fix the issue in `to_timestamp` after RC2 passes?

**Comment by MaxGekk** [comment]
> maybe we can just fix the issue in to_timestamp after RC2 passes?

@cloud-fan I have tried to port all functions on `TimestampParser` by rewriting `DateTimeUtils.newDateFormat`:
```scala
  def newDateFormat(formatString: String, timeZone: TimeZone): TimestampParser = {
    new TimestampParser(FastDateFormat.getInstance(formatString, timeZone))
  }
```
from
```scala
  def newDateFormat(formatString: String, timeZone: TimeZone): DateFormat = {
    val sdf = new SimpleDateFormat(formatString, Locale.US)
    sdf.setTimeZone(timeZone)
    // Enable strict parsing, if the input date/format is invalid, it will throw an exception.
    // e.g. to parse invalid date '2016-13-12', or '2016-01-12' with  invalid format 'yyyy-aa-dd',
    // an exception will be throwed.
    sdf.setLenient(false)
    sdf
  }
```
but I got a few test failures due to strict mode set via **`sdf.setLenient(false)`**. I haven't found a way to set similar mode for `FastDateFormat`.

**Comment by cloud-fan** [comment]
OK then this doc change LGTM

**Comment by dongjoon-hyun** [comment]
Thank you for the investigation, @MaxGekk . And, thank you for the conclusion, @cloud-fan .

**Comment by MaxGekk** [comment]
@dongjoon-hyun Kindly remind you about this minor fix.

**Comment by dongjoon-hyun** [comment]
Sure, @MaxGekk . Sorry for making you wait. Merged to `branch-2.4`.

**Pull request closed without merging**
